Engine/DataFeeds/ApiDataFileProvider.cs

```diff
@@ -0,0 +1,60 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License"); 
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using QuantConnect.Configuration;
+using QuantConnect.Interfaces;
+using QuantConnect.Logging;
+
+namespace QuantConnect.Lean.Engine.DataFeeds
+{
+    /// <summary>
+    /// Attempts to download data from the api
+    /// </summary>
+    public class ApiDataFileProvider : IDataFileProvider
+    {
+        private readonly int _uid = Config.GetInt("job-user-id", 0);
+        private readonly string _token = Config.Get("api-access-token", "1");
+        private readonly string _dataPath = Config.Get("data-folder", "../../../Data/");
+        public bool Fetch(Symbol symbol, Resolution resolution, DateTime date)
+        {
+            Log.Trace(
+                string.Format(
+                    "Attempting to get data from QuantConnect.com's data library for symbol({0}), resolution({1}) and date({2}).",
+                    symbol.ID, resolution, date.Date.ToShortDateString()));
+
+            var api = new Api.Api();
+            api.Initialize(_uid, _token, _dataPath);
+
+            var download = api.DownloadData(symbol, resolution, date);
```

Comment:
Hi guys, I just found this cool work!  I was in the middle of writing one on my own but am happy to see an interface.  I would suggest that if the file already exists, the backtester would just read that during backtesting.  Otherwise, it would fetch the IFileDataProvider and write that to disk in Lean format after feeding to the algorithm.  Perhaps it will be extensible by the consumer to do this too.
Either way you decide, I'm excited to see this new feature so I can pipe other data sources quickly.